internal/pkg/composable/providers/kubernetes/watcher.go

```diff
@@ -0,0 +1,285 @@
+				// However, there is no guarantee that all events from a watch are processed by the receiver. To ensure that missed events
+				// are properly handled, a period re-list is done to ensure that every state within the system is effectively handled.
+				// In this case, we are making sure that we are enqueueing an "add" event because, an runner that is already in Running
+				// state should just be deduped by autodiscover and not stop/started periodically as would be the case with an update.
+				w.enqueue(n, add)
+			}
+
+			//We check the type of resource and only if it is namespace or node return the cacheObject
+			switch resource.(type) {
+			case *Namespace:
+				w.cacheObject(o)
+			case *Node:
+				w.cacheObject(o)
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if transformFunc != nil {
+		err = informer.SetTransform(transformFunc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return w, nil
+}
+
+// AddEventHandler adds a resource handler to process each request that is coming into the watcher
+func (w *watcher) AddEventHandler(h ResourceEventHandler) {
+	w.handler = h
+}
+
+// GetEventHandler returns the watcher's event handler
+func (w *watcher) GetEventHandler() ResourceEventHandler {
+	return w.handler
+}
+
+// Store returns the store object for the resource that is being watched
+func (w *watcher) Store() cache.Store {
+	return w.store
+}
+
+// Client returns the kubernetes client object used by the watcher
+func (w *watcher) Client() kubernetes.Interface {
+	return w.client
+}
+
+// CachedObject returns the old object in cache during the last updated event
+func (w *watcher) CachedObject() runtime.Object {
+	return w.cachedObject
+}
+
+// Start watching pods
+func (w *watcher) Start() error {
+	go w.informer.Run(w.ctx.Done())
+
+	if !cache.WaitForCacheSync(w.ctx.Done(), w.informer.HasSynced) {
+		return fmt.Errorf("kubernetes informer unable to sync cache")
+	}
+
+	w.logger.Debugf("cache sync done")
+
+	// Wrap the process function with wait.Until so that if the controller crashes, it starts up again after a second.
+	go wait.Until(func() {
+		for w.process(w.ctx) {
+		}
+	}, time.Second*1, w.ctx.Done())
+
+	return nil
+}
+
+func (w *watcher) Stop() {
+	w.queue.ShutDown()
+	w.stop()
+}
+
+// enqueue takes the most recent object that was received, figures out the namespace/name of the object
+// and adds it to the work queue for processing.
+func (w *watcher) enqueue(obj interface{}, state string) {
+	// DeletionHandlingMetaNamespaceKeyFunc that we get a key only if the resource's state is not Unknown.
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	if deleted, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		w.logger.Debugf("Enqueued DeletedFinalStateUnknown contained object: %+v", deleted.Obj)
+		obj = deleted.Obj
+	}
+	w.queue.Add(&item{key, obj, state})
```

Comment:
Adding a comment here but can be it's own issue/PR. 
According to https://github.com/topfreegames/maestro/pull/473
https://github.com/kubernetes/kubernetes/issues/103789
https://github.com/kubernetes/sample-controller/blob/master/docs/controller-client-go.md
https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go
we're supposed to only put the `key` on the work queue, and retrieve it from the indexer when processing it. Here we're storing the object and later retrieving it by key, so we're storing extra copies.

Reply:
We should file that against https://github.com/elastic/elastic-agent-autodiscover. To be honest, I'm not convinced this queue is even necessary, but in this PR I just wanted to make the smallest amount of changes necessary.

Reply:
https://github.com/elastic/elastic-agent-autodiscover/issues/108